src/peft/tuners/lora/variants.py

```diff
@@ -315,3 +317,162 @@ class DoraConv3dVariant(_DoraConvNdVariant):
     def init(module: Conv3d, adapter_name: str, **kwargs: Any) -> None:
         dora_layer = DoraConv3dLayer(fan_in_fan_out=False)
         _DoraConvNdVariant.init_convd_variant(module, adapter_name, dora_layer=dora_layer)
+
+
+class SineLoraLinearVariant(LoraVariant):
+    @staticmethod
+    def init(module: Linear, adapter_name: str, **kwargs) -> None:
+        module.sinelora_frequency = kwargs["sinelora_frequency"]
+
+        module.sinelora_scaling = kwargs["sinelora_scaling"]
+        if module.sinelora_scaling is None:
+            module.sinelora_scaling = math.sqrt(module.in_features)
+
+    @staticmethod
+    def unmerge(module: Linear, active_adapter: str, orig_weight: torch.Tensor) -> torch.Tensor:
+        orig_dtype = orig_weight.dtype
+        lora_A = module.lora_A[active_adapter]
+        lora_B = module.lora_B[active_adapter]
+        lora_scaling = module.scaling[active_adapter]
+        delta_weight = (
+            torch.sin(module.sinelora_frequency * lora_A.weight.T @ lora_B.weight.T).T
+            / module.sinelora_scaling
+            * lora_scaling
+        )
+
+        delta_weight = delta_weight.to(orig_dtype)
+        unmerged_weight = orig_weight - delta_weight
+        return unmerged_weight
+    
+    
+    @staticmethod
+    def merge_safe(module: Linear, active_adapter: str, orig_weight: torch.Tensor) -> torch.Tensor:
+        orig_dtype = orig_weight.dtype
+        lora_A = module.lora_A[active_adapter]
+        lora_B = module.lora_B[active_adapter]
+        lora_scaling = module.scaling[active_adapter]
+        delta_weight = (
+            torch.sin(module.sinelora_frequency * (lora_A.weight.T @ lora_B.weight.T)).T
+            / module.sinelora_scaling
+            * lora_scaling
+        )
+        merged_weight = orig_weight + delta_weight
+        if not torch.isfinite(merged_weight).all():
+            raise ValueError(f"NaNs detected in merged weights for adapter {active_adapter}")
+        module._cache_store(f"{active_adapter}-delta_weight", delta_weight)
```

Comment:
I don't think we need to cache the delta weights?

The same goes for `merge_unsafe` and the `Embedding` implementation.